### PR TITLE
Upgrade control plane before machine deployment

### PR DESCRIPTION
**Pull request by jiayiwang7**
*Issue #, if available:*

https://github.com/aws/eks-anywhere/issues/100

*Description of changes:*

- Break provider CAPI template into template-cp and template-md
- Return CAPI control plane and machine deployment as bytes in provider GenerateClusterAPI code for both create and upgrade
- Kubectl apply CAPI bytes instead of actual files in cluster_manager
- For upgrade, apply CAPI control plane template, wait for all control planes successfully upgrade, and then apply machine deployment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/clustermanager/cluster_manager.go
Comment (jiayiwang7) [inline pkg/clustermanager/cluster_manager.go:264]: still tweaking the wait logic for upgrade. WIP

**Review by mdsgabriel — APPROVED** [review]

File: pkg/clustermanager/cluster_manager.go
Comment (mrajashree) [inline pkg/clustermanager/cluster_manager.go:247]: since the mdContent is appended to cpContent, will the kubectl apply command apply the updated spec for cp and md at the same time? should we first wait for controlplane upgrade to go through and then apply mdContent separately?

Comment (jiayiwang7) [reply pkg/clustermanager/cluster_manager.go:247]: yea that's what i am doing. the mdContent is only appended to cp for writing the spec file to local. we still apply cpContent first, then mdContent.

Comment (mrajashree) [reply pkg/clustermanager/cluster_manager.go:247]: ah gotcha, I thought append yaml resources on line 241 was appending mdContent to cpContent so cpContent is getting updated, but that's not the case

File: pkg/clustermanager/cluster_manager.go
Comment (mrajashree) [inline pkg/clustermanager/cluster_manager.go:254]: I think these checks are already being done in this function [line 275](https://github.com/aws/eks-anywhere/pull/264/files#diff-2a3001a4bccb1fa62c0978f41ec1a89279d6e8cbc47e531f59c754b5df5715b6R275) onwards

File: pkg/clustermanager/cluster_manager.go
Comment (mrajashree) [inline pkg/clustermanager/cluster_manager.go:271]: do you think we should modify this to only check the nodeRef and nodeHealthy for controlplane nodes at this point? and then after applying mdContent run this check for worker nodes/all nodes at that point

File: pkg/clustermanager/cluster_manager.go
Comment (jiayiwang7) [reply pkg/clustermanager/cluster_manager.go:254]: this is a confusing part for me too. so the original code actually check the etcd status twice. i thought we did it on purpose.

File: pkg/clustermanager/cluster_manager.go
Comment (jiayiwang7) [reply pkg/clustermanager/cluster_manager.go:271]: that makes sense to me. let me test it to see if it does the work

File: pkg/clustermanager/cluster_manager.go
Comment (jiayiwang7) [reply pkg/clustermanager/cluster_manager.go:254]: chatted with rajashree offline, it seems double check etcd ready status is not necessary. gonna remove one

File: pkg/providers/vsphere/config/template-cp.yaml
Comment (vivek-koppuru) [inline pkg/providers/vsphere/config/template-cp.yaml:90]: So `Cluster` is considered part of CP? I am wondering if we want to template some of that info out too, or change cp to `Core` or something.

Comment (jiayiwang7) [reply pkg/providers/vsphere/config/template-cp.yaml:90]: I thought about it once but `Core` feels confusing to me a bit. CP might not be the best term but it is less confusing.

Comment (vivek-koppuru) [reply pkg/providers/vsphere/config/template-cp.yaml:90]: I think the ideal separation is to have three, cp, worker, and core/cluster. But I think that's a lot more scope imo. I think of core because it includes external etcd as well and is needed to operate a kubernetes cluster, but yea I get the confusion.

File: controllers/controllers/resource/template.go
Comment (jiayiwang7) [reply controllers/controllers/resource/template.go:79]: The function is `GenerateCapiSpecMD`, it's mainly about generating CAPI machine deployment. CAPI does not have worker node groups. I feel it is more proper to keep the MD in this case.

Comment (vivek-koppuru) [reply controllers/controllers/resource/template.go]: I guess the way I thought about it was "generate all cluster api spec that is related to worker nodes" rather than specifically for machine deployments. Cp makes sense above because control plane is more generic and is something that we follow too. By naming it worker, we are just letting this function contain all the context as to what constitutes a worker template, which in our case is MachineDeployment related templates and anything else we introduce in the future.

Comment (jiayiwang7) [reply controllers/controllers/resource/template.go]: The way I think abt it is the function itself is generating CAPI spec. It does not contain our EKS-A business logic. Also all the configs in the template-md are CAPI MD related.

File: pkg/clustermanager/cluster_manager.go
Comment (g-gaston) [inline pkg/clustermanager/cluster_manager.go:157]: curious, what is this for?

File: controllers/controllers/resource/template.go
Comment (g-gaston) [inline controllers/controllers/resource/template.go:79]: nit: I think it should be `CAPI` instead of `Capi`

**Review by g-gaston — COMMENTED** [review]
Looks good, no mayor concerns from my side

File: pkg/clustermanager/cluster_manager.go
Comment (jiayiwang7) [reply pkg/clustermanager/cluster_manager.go:157]: since we removed the write from provider code, this is the only place to write CAPI spec to local

File: pkg/providers/vsphere/testdata/expected_results_main_no_machinetemplate_update_cp.yaml
Comment (jiayiwang7) [reply pkg/providers/vsphere/testdata/expected_results_main_no_machinetemplate_update_cp.yaml:352]: yes. the MD still has the `infrastructureRef` that points to the old `machinetemplate`. the name "no_machinetemplate_update_cp" is sufficient.

File: pkg/providers/vsphere/vsphere_test.go
Comment (jiayiwang7) [reply pkg/providers/vsphere/vsphere_test.go:268]: it's not removed. it is just moved to another test function. see line 313 `TestProviderGenerateClusterApiSpecForUpgradeOIDC` bc for OIDC we only need to test CP spec.

File: pkg/providers/provider.go
Comment (jiayiwang7) [reply pkg/providers/provider.go]: if we dont use acronyms, i prefer to call it `WorkerNodes` (align w our API) instead of introducing a new name. @vivek-koppuru thoughts?

Comment (vivek-koppuru) [reply pkg/providers/provider.go]: Agreed, fine with either

File: controllers/controllers/resource/template.go
Comment (jiayiwang7) [reply controllers/controllers/resource/template.go:79]: i can make the change across the board

File: pkg/providers/vsphere/vsphere.go
Comment (jiayiwang7) [reply pkg/providers/vsphere/vsphere.go:1185]: i didnt write the code. i just moved it. but i can update that

**Review by danbudris — APPROVED** [review]

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-762276747" title="Approved">danbudris</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#" title="Author self-approved">jiayiwang7</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-760084827" title="Approved">mdsgabriel</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [danbudris,jiayiwang7,mdsgabriel]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: pkg/providers/vsphere/vsphere.go
Comment (vivek-koppuru) [inline pkg/providers/vsphere/vsphere.go:1084]: I forget, does this make the capacity 4 or the size 4 with 4 `nil` appended to it? I ran into this problem before so wanted to make sure.

**Comment by abhay-krishna** [comment]
@jiayiwang7: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-cluster-controller-tooling-presubmit | 5b29e4f7a64931edab51f2987db137dd3a70c8bc | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/264/eks-anywhere-cluster-controller-tooling-presubmit/1441088083175936000) | `/test eks-anywhere-cluster-controller-tooling-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=264). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:jiayiwang7).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Review by vivek-koppuru — APPROVED** [review]
/lgtm

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-762276747" title="Approved">danbudris</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#" title="Author self-approved">jiayiwang7</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-760084827" title="Approved">mdsgabriel</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-762302312" title="LGTM">vivek-koppuru</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [danbudris,jiayiwang7,mdsgabriel,vivek-koppuru]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by abhay-krishna** [comment]
@jiayiwang7: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-cluster-controller-tooling-presubmit | 7e6f7a33f02bec52372f07e07028f421aa84f5ff | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/264/eks-anywhere-cluster-controller-tooling-presubmit/1441089702819008512) | `/test eks-anywhere-cluster-controller-tooling-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=264). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:jiayiwang7).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
@jiayiwang7: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-cluster-controller-tooling-presubmit | 7e6f7a33f02bec52372f07e07028f421aa84f5ff | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/264/eks-anywhere-cluster-controller-tooling-presubmit/1441096036952051712) | `/test eks-anywhere-cluster-controller-tooling-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=264). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:jiayiwang7).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Review by mrajashree — APPROVED** [review]

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-762276747" title="Approved">danbudris</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#" title="Author self-approved">jiayiwang7</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-760084827" title="Approved">mdsgabriel</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-762358083" title="Approved">mrajashree</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-762302312" title="LGTM">vivek-koppuru</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [danbudris,jiayiwang7,mdsgabriel,mrajashree,vivek-koppuru]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: pkg/types/resources.go
Comment (g-gaston) [inline pkg/types/resources.go:23]: I don't think this method is doing what the name is saying it does
`HasLabels` means to me: "this machine has all these labels"
maybe something like `HasAnyLabel`?
btw little old unit test would be great

File: pkg/providers/vsphere/vsphere.go
Comment (g-gaston) [inline pkg/providers/vsphere/vsphere.go]: Why 4?
shouldn't it be the addition of all cidr blocks and noproxy lengths +1?
Also, when using append, I always find safer/cleaner using `make([]string, 0, 4)`. That would make sure the code doesn't break if for whatever reason we end up appending less that 4 items

File: pkg/providers/vsphere/vsphere.go
Comment (g-gaston) [inline pkg/providers/vsphere/vsphere.go]: same question here about 4 capacity

Comment (jiayiwang7) [reply pkg/providers/vsphere/vsphere.go]: it is already updated in latest code...

**Comment by abhay-krishna** [comment]
New changes are detected. LGTM label has been removed.

**Comment by abhay-krishna** [comment]
@jiayiwang7: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-cluster-controller-tooling-presubmit | a2922029781a72a9c060446695c92c5b72d4375f | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/264/eks-anywhere-cluster-controller-tooling-presubmit/1441151043168636928) | `/test eks-anywhere-cluster-controller-tooling-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=264). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:jiayiwang7).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Review by g-gaston — APPROVED** [review]
/lgtm

**Comment by abhay-krishna** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-762276747" title="Approved">danbudris</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-762520919" title="LGTM">g-gaston</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#" title="Author self-approved">jiayiwang7</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-760084827" title="Approved">mdsgabriel</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-762358083" title="Approved">mrajashree</a>*, *<a href="https://github.com/aws/eks-anywhere/pull/264#pullrequestreview-762302312" title="LGTM">vivek-koppuru</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=aws%2Feks-anywhere).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/aws/eks-anywhere/blob/main/OWNERS)~~ [danbudris,g-gaston,jiayiwang7,mdsgabriel,mrajashree,vivek-koppuru]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by abhay-krishna** [comment]
@jiayiwang7: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-cluster-controller-tooling-presubmit | a2922029781a72a9c060446695c92c5b72d4375f | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/264/eks-anywhere-cluster-controller-tooling-presubmit/1441159471794491392) | `/test eks-anywhere-cluster-controller-tooling-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=264). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:jiayiwang7).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by abhay-krishna** [comment]
New changes are detected. LGTM label has been removed.

**Comment by abhay-krishna** [comment]
@jiayiwang7: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
eks-anywhere-cluster-controller-tooling-presubmit | d3f15fe394c6522ca28a45b6a9af22b581c42df0 | [link](https://prow.eks.amazonaws.com/view/s3/prowpresubmitsdataclusterstack-prowbucket7c73355c-vfwwxd2eb4gp/pr-logs/pull/aws_eks-anywhere/264/eks-anywhere-cluster-controller-tooling-presubmit/1441164462525517824) | `/test eks-anywhere-cluster-controller-tooling-presubmit`

[Full PR test history](https://prow.eks.amazonaws.com/pr-history?org=aws&repo=eks-anywhere&pr=264). [Your PR dashboard](https://prow.eks.amazonaws.com/pr?query=is:pr+state:open+author:jiayiwang7).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Pull request merged**
